### ISSUE TITLE
RFC: How components evolve independently

### ISSUE THREAD

**Issue by dzearing** [OfficeDev/office-ui-fabric-react#5450]
The `Persona` is a good example of a "heavy component". There are a lot of wrapper DIVs generated for something that should be very simple. We want to refactor it but there are things we consider breaking:

* The DOM structure changes (it simplifies)
* The class names change

While we can make props contracts be backwards compatible with deprecation usage, we still consider DOM and class name changes as breaking. A partner may have a css selector which relies on the DOM structure or class names. But they're great changes as they reduce DOM weight! We want a safe way to evolve these changes so that we minimize the impact on partners.

Ideally, we'd do this in a major release, but it can't happen overnight, and we learned from the last major release that long running major release branches can easily fall behind master and create a large tax in keeping updated.

Goals:

* Components can change DOM structure and class names
* Users who want the new component can opt into it before a major release happens.
* Components that have "next" versions are easy to merge into a major release.

## Proposal 1: Subdirectories

In the Persona component folder, we create a `next` folder, with a snapshot of the Persona code and use that for the "next major release" refactor. Partners can use webpack aliasing to remap old persona index to new index. Basically we wouldn't export new Persona by default, but moving to the next major fabric release would mean moving all "next" folders to current.

This is a pretty simple pattern to implement and very transparent to partners, though completely non-discoverable which means it won't get as tested.

## Proposal 2: Micro packages + subdirectories

As an improvement on the first proposal, we create `@uifabric/persona` package which contains the current Persona and "next" persona code. We run a separate publishing workflow for the "next" release, so that people can either npm install the latest or the "next" version. The `office-ui-fabric-react` package would depend on this version (and re-export it) with a range of `>=6 <=7`, enabling semver to be met if the consumer opts to use the "next" version of the persona package. (This may need some discussion, I'm a little unsure of the semver range.)

It's also possible to  public an `@next` version of the oufr package, which re-exports all `@next` sub-packages.

## Proposal 3: Use `@uifabric/experiments` package

New versions of components go in the experiments package.  This could work, but you'd still need to use webpack aliasing to test the new code. And, we've seen issues with the experiments package.

## Proposal 4: global class names and DOM structure are not a contract

I think I'd be OK assuming adding DOM structure is not a contract, but removing any existing class name or changing it will break visuals. I do think in a future release we will want to move away from global class names once all things can be customized using `styles` props. But not now.

**Comment by micahgodbolt** [comment]
I'm less of a fan of experiments. We'd still have the drifting apart problem as features are added to old persona without being addressed in the new one. It'd be much easier to ensure that the two components stay in sync if they are in the same folder....or as you mention, in the same micro package. Micropackages are cool, as long as you feel that it'd scale properly. 

As for #4, I think this is still a good idea for components that have been converted to css-in-js. Maybe send out a deprecation warning now, or with 7.0, stating that DOM and existing Classnames can change without a major release soon, and to move all custom styling to styles prop.

**Comment by dzearing** [comment]
I think a separate topic, but very related, is how we coordinate major bumps. Major bumps give us the opportunity to remove deprecated things and to change DOM and potentially classname structure. But we want to do it very incrementally and rarely. We want things to be reliable and consistent, but clearly fixing bugs sometimes means breaking changes.

We learned two things from our 6.0 release:

1. It is bad to bump things artificially. We bumped merge-styles, even though nothing major changed, which means that moving to the next component major release requires a new version of merge-styles.  We need to decouple the components from the foundation. The foundational things should almost never change, but the components can.

2. Major release branches are the devil and should be very short lived. That means that current branch (master) needs the ability to contain both old and new, in a very easily identifiable way, such that moving from old to new is a few minutes operation to manage, and that the "next major" branch can live only for a few weeks (the life of the beta rollout.)
